doc/workshops/plugin-development.rst

```diff
@@ -2,4 +2,917 @@
+
+
+Timer updates
+~~~~~~~~~~~~~
+
+The first complaint, naturally, is that our timer is static. To activate it we can use the ``QTimer`` of ``PomodoroTimerStatus`` as an instance of ``BaseTimerStatus``.
+
+Let's go back to ``widgets.py`` and add this constant below the import lines (line 22).
+
+.. code-block:: python
+
+   # --- Constants
+   # ------ Time limits by default
+
+   POMODORO_DEFAULT = 25 * 60 * 1000  # 25 mins in milliseconds
+
+In the ``__init__`` method of ``PomodoroTimerStatus`` we are going to substitute:
+
+.. code-block:: python
+
+       self.value = "25:00"
+
+by
+
+.. code-block:: python
+
+       self.value = ""
+
+       # Actual time limits
+       self.pomodoro_limit = POMODORO_DEFAULT
+       self.countdown = self.pomodoro_limit
+
+       self.timer.setInterval(1000)
+       self.timer.timeout.connect(self.update_timer)
+       self.timer.start()
+
+So far we have created a default value (``POMODORO_DEFAULT``) for the timer duration during pomodoros.
+We have added it to the ``pomodoro_limit`` field to be able to configure it.
+And we initialize with that value the ``countdown`` field that will be modified over time.
+Since the default interval of the ``BaseTimerStatus`` timer is 2 seconds we set it to 1 second (1000 in milliseconds).
+
+The function of the timer is to activate functions periodically, this is done through the method ``timeout.connect()`` to which we pass as parameter the reference to the ``update_timer`` function that will perform the required adjustments.
+
+Let's implement ``update_timer`` at the end of the file:
+
+.. code-block:: python
+
+       def display_time(self, time):
+           """Remaining time in minutes:seconds format."""
+
+           minutes = int((time / (1000 * 60)) % 60)
+           seconds = int((time / 1000) % 60)
+           return f"{minutes:02d}:{seconds:02d}"
+
+       def update_timer(self):
+           """Updates the countdown until it reaches zero."""
+
+           remaining_time = self.display_time(self.countdown)
+
+           if remaining_time == "00:00":
+               self.value = remaining_time
+           else:
+               # Update the current timer by decreasing the current running time by one second
+               self.countdown -= 1000
+               self.value = remaining_time
+
+Here we rely on the ``display_time`` method that converts the current ``countdown`` value, which is measured in milliseconds, into a human-readable format.
+
+And ``update_timer`` simply keeps updating the countdown until we reach zero.
+
+If we run Spyder again we will find that our timer has come to life.
+
+.. image:: images/pd_timer_countdown.gif
+   :alt: Timer countdown working.
+
+
+Timer controls
+~~~~~~~~~~~~~~
+
+Now we need a way to control our timer. It may not be the best option, but we can add some buttons to the toolbar to show how to add controls and actions to Spyder.
```

Comment:
Not sure why we need to mention that it may not be the best option.

Reply:
In this case the idea was to leave these plugin design considerations as homework for the participants. 
It is not the best option, but perhaps the simplest and the one that serves to illustrate how to add buttons to the toolbar.